### PR TITLE
Sequence Operator Improvements

### PR DESCRIPTION
**Pull request by sbodenstein**
## Description ##
This PR does the following:
- Adds an optional argument `axis` to `SequenceLast`, `SequenceMask` and `SequenceReverse`. The default is `axis=0`. Values for `axis=1` and `axis=0` are implemented for `SequenceLast` and `SequenceMask`. 
- Rigorous tests are added for all sequence ops. 
- Bug in `SequenceMask` was fixed (incorrect behaviour for `kAddTo`).
- Performance improvements: `SequenceLast` avoids any movement of data to the CPU, and uses a much faster implementation similar to `pick`. `SequenceMask` now supports inplace.

Axis support will be added to `SequenceReverse` in a separate PR.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Feature1, tests, (and when applicable, API doc)
- [x] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/sequence_last-inl.h [hunk 1]
```diff
@@ -45,16 +45,46 @@ namespace op {
 namespace seq_last {
 enum SequenceLastOpInputs { kData, kSequenceLength };
 enum SequenceLastOpOutputs { kOut };
+enum SequenceLastOpResource { kTempSpace };
 }
 
 struct SequenceLastParam : public dmlc::Parameter<SequenceLastParam> {
   bool use_sequence_length;
+  uint32_t axis;
```
Comment (reminisce) [inline src/operator/sequence_last-inl.h]: Is it possible to support negative axis in the future? If so, it's better to define it as a signed integer.

File: src/operator/sequence_last-inl.h [hunk 2]
```diff
@@ -184,17 +252,22 @@ class SequenceLastProp : public OperatorProperty {
     CHECK_EQ(in_shape->size(), param_.use_sequence_length ? 2U : 1U)
         << "Input:[data, sequence_length]";
 
+    if ((param_.axis != 0) && (param_.axis != 1)) {
```
Comment (reminisce) [inline src/operator/sequence_last-inl.h]: You can simplify this by using CHECK.

File: src/operator/sequence_reverse-inl.h [hunk 3]
```diff
@@ -221,6 +224,10 @@ class SequenceReverseProp : public OperatorProperty {
     CHECK_EQ(in_shape->size(), param_.use_sequence_length ? 2U : 1U)
         << "Input:[data, sequence_length]";
 
+    if ((param_.axis != 0)) {
```
Comment (reminisce) [inline src/operator/sequence_reverse-inl.h]: Use CHECK_EQ to simplify the code.

File: src/operator/sequence_last-inl.h (hunk 1 above)
Comment (sbodenstein) [reply src/operator/sequence_last-inl.h]: Yes. Fixed, thanks.

File: src/operator/sequence_last-inl.h [hunk 4]
```diff
@@ -45,16 +45,46 @@ namespace op {
...
   DMLC_DECLARE_PARAMETER(SequenceLastParam) {
     DMLC_DECLARE_FIELD(use_sequence_length)
         .set_default(false)
         .describe(
-            "If set to true, this layer takes in an extra input parameter `sequence_length` "
+            "If set to true, this layer takes in an extra input parameter "
+            "`sequence_length` "
             "to specify variable length sequence");
+    DMLC_DECLARE_FIELD(axis).set_default(0).describe(
+        "The sequence axis. Only values of 0 and 1 are current supported.");
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply src/operator/sequence_last-inl.h]: fixed.

File: src/operator/sequence_last-inl.h [hunk 5]
```diff
@@ -45,16 +45,46 @@ namespace op {
...
+};
+
+template <int req>
+struct SequenceLastKernel {
+  template <typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType *out, const DType *in,
+                                  const DType *idx, int offset1, int offset2,
+                                  mshadow::Shape<2> oshape) {
+    auto opos = mxnet_op::unravel(i, oshape);
+    int seqpos = static_cast<int>(idx[opos[0]]) - 1;
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply src/operator/sequence_last-inl.h]: Done.

File: src/operator/sequence_last-inl.h [hunk 6]
```diff
@@ -74,33 +145,32 @@ class SequenceLastOp : public Operator {
     CHECK_EQ(out_data.size(), 1U);
     Stream<xpu> *s = ctx.get_stream<xpu>();
 
+    // only support axis of 0 or 1 for now
+    bool axis = static_cast<bool>(param_.axis);
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply src/operator/sequence_last-inl.h]: Agreed. Fixed.

File: src/operator/sequence_last-inl.h (hunk 2 above)
Comment (sbodenstein) [reply src/operator/sequence_last-inl.h]: Done.

File: src/operator/sequence_reverse-inl.h (hunk 3 above)
Comment (sbodenstein) [reply src/operator/sequence_reverse-inl.h]: Done.

**Comment by sbodenstein** [comment]
@reminisce: can this be merged?

File: tests/python/unittest/test_operator.py [hunk 7]
```diff
@@ -2311,40 +2311,86 @@ def test_l2_normalization():
...
+# Numpy Implementation of Sequence Ops
+def sequence_last_numpy(array, lengths, axis):
+    # create new array of dims [batch, seqlen, ...]
+    array2 = np.moveaxis(array, axis, 1)
+    dims = array2.shape
+    if lengths is None:
+        return array2[:, -1]
+    lengths = list(lengths)
+    return np.array([array2[i, int(lengths[i]) - 1] for i in range(dims[0])])
+
```
[the remark this answers is not in the document]
Comment (sbodenstein) [reply tests/python/unittest/test_operator.py]: Fixed.

**Comment by asmushetzel** [comment]
I just accidentally stumbled upon the bug with the kAddTo in SequenceMask (on a slightly older codebase). Nice to see it already fixed ;-)

**Pull request merged**